Implement remainingUnionDebt and supplyUnionCredits

mheinzel opened the pull request:
There is an issue in the lockstep tests with how `SupplyPortionOfDebt` works, so I had to disable that action for now. I'll see if we can salvage it, but we'd probably have to relax the `Eq (Obs h a)` instance for `UnionCredits`. Ideally we'd have separate cases for `SupplyUnionCredits` and `SupplyPortionOfDebt` to enforce that for one the implementation can't return fewer and for the other it can't return more leftovers than the model.

I'm also still adding a property based test for supplying into the `MergingTree`.

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -154,19 +157,39 @@ new ::
-        MergeMaybeCompleted
-        (OngoingMerge runs merge)
+      let dupRun r = withRollback reg (dupRef r) releaseRef
+      case V.length nonEmptyRuns of
+        0 -> do
+          r <- dupRun (V.head inputRuns)
+          unsafeNew
+            (NumRuns (V.length inputRuns))  -- TODO: is this sensible?
+            (MergeDebt 0)
+            (SpentCredits 0)
+            MergeKnownCompleted
+            (CompletedMerge r)
+        1 -> do
+          r <- dupRun (V.head nonEmptyRuns)
+          unsafeNew
+            (NumRuns (V.length inputRuns))  -- TODO: is this sensible?
+            (MergeDebt 0)
+            (SpentCredits 0)
+            MergeKnownCompleted
+            (CompletedMerge r)
+        _ -> do
+          mergeRuns <- V.mapM dupRun nonEmptyRuns
+          merge <- fromMaybe (error "newMerge: merges can not be empty")
+            <$> Merge.new hfs hbio runParams ty resolve runPaths mergeRuns
+          unsafeNew
+            (NumRuns (V.length inputRuns))  -- TODO: is this sensible?
+            (numEntriesToMergeDebt (V.foldMap' Run.size mergeRuns))
+            (SpentCredits 0)
+            MergeMaybeCompleted
+            (OngoingMerge mergeRuns merge)
```
I think it's a good idea to leave out empty runs and directly creating completed merges when they are trivial. However, it's a bit ambiguous what the number of runs for such a merge should be. Always using the number of input runs probably has the least number of surprises, even if that doesn't match the number of runs we hold references for?

mheinzel replied:
Resolved in #635

mheinzel commented on src/Database/LSMTree/Internal/Config.hs:
```diff
@@ -95,6 +97,16 @@ defaultTableConfig =
       , confMergeSchedule     = defaultMergeSchedule
       }
 
+data RunLevelNo = RegularLevel LevelNo | UnionLevel
+
+runParamsForLevel :: TableConfig -> RunLevelNo -> RunParams
+runParamsForLevel conf@TableConfig {..} levelNo =
+    RunParams
+      { runParamCaching = diskCachePolicyForLevel confDiskCachePolicy levelNo
+      , runParamAlloc   = bloomFilterAllocForLevel conf levelNo
+      , runParamIndex   = indexTypeForRun confFencePointerIndex
+      }
```
I define this in one place for both regular and union levels. Does that seem sensible?

dcoutts commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -272,3 +295,42 @@ finalise mergeState = releaseMTS =<< readMVar mergeState
+        PreExistingMergingRun mr -> MR.remainingMergeDebt mr
+
+mergeDebt :: Foldable f => f (MergeDebt, NumEntries) -> (MergeDebt, NumEntries)
+mergeDebt debts =
```
might be better to call this `sumMergeDebt` or something. Otherwise it can be confused with being the same concept as `MergeDebt`.

mheinzel replied:
`sum` doesn't quite feel right, because it's not just adding up the debts and sizes of the child merges, it also includes the debt of the new merge. So I called it `debtOfNestedMerge` now.

dcoutts commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -781,6 +783,34 @@ remainingMergeDebt (DeRef mr) = do
+-- | Helper function to assert common invariants for functions that supply
+-- credits.
+supplyChecked ::
+     forall m r s. (HasCallStack, Monad m)
```
This constraint is redundant when `!NO_IGNORE_ASSERTS` and generates a warning.

mheinzel replied:
Yeah, fixed now, but it's a recurring annoying issue.

dcoutts commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -789,27 +819,24 @@ remainingMergeDebt (DeRef mr) = do
-    assert (suppliedCredits' == mergeDebtAsCredits (totalMergeDebt mr)
-            || leftoverCredits == 0) $
-      pure r
+  -> m MergeCredits
```
The reason we were returning all three `MergeCredits` values was so that in the `MergeSchedule` we would have the opportunity to trace the before/after credits. I still think that'd be useful, though we've not done it yet.

dcoutts replied:
Oh I see, right. I'd forgotten about the absolute/relative and which ones were used where. Yes that's fine.

dcoutts commented on bench/micro/Bench/Database/LSMTree/Internal/Index.hs:
```diff
@@ -7,8 +7,10 @@ import           Control.DeepSeq (rnf)
 import           Criterion.Main (Benchmark, Benchmarkable, bench, bgroup, env,
                      whnf)
-#if __GLASGOW_HASKELL__ < 910
+#if MIN_VERSION_base(4,20,0)
+#else
```
```
#if !MIN_VERSION_base(4,20,0)
...
#endif
```

mheinzel commented on test/Test/Database/LSMTree/StateMachine.hs:
```diff
@@ -1983,30 +1983,21 @@ arbitraryActionWithVars _ label ctx (ModelState st _stats) =
-        , isJust (eqT @h @ModelIO.Table)
+          -- TODO: the model always has remaining debt 0, so it will only supply
+          -- 0, so it always has leftovers 0. This doesn't match the real impl.
+        , False
```
Resolved in #636

dcoutts reviewed (commented):
Partial review. I've only got as far as "implement supplyUnionCredits" (started but not finished that patch).

dcoutts commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -327,10 +353,170 @@ remainingMergeDebtPendingMerge (PendingMerge _ prs mts) = do
+    expectCompletedChildren reg (PendingMerge ty prs mts) = do
+        rs1 <- V.forM prs $ \case
+          PreExistingRun r -> do
+            delayedCommit reg (releaseRef r)  -- only released at the end
+            withRollback reg (dupRef r) releaseRef
```
Why do we duplicate the ref and also release the old one? Why not just reuse the one reference directly?

mheinzel replied:
Makes it more uniform (`MR.expectCompleted` below makes a new reference) and if we forget to release it then error message gives us more helpful location info.

dcoutts commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -327,10 +353,170 @@ remainingMergeDebtPendingMerge (PendingMerge _ prs mts) = do
+            delayedCommit reg (releaseRef r)  -- only released at the end
+            withRollback reg (dupRef r) releaseRef
+          PreExistingMergingRun mr -> do
+            delayedCommit reg (releaseRef mr)  -- only released at the end
+            withRollback reg (MR.expectCompleted mr) releaseRef
```
Whereas here it makes sense because we're allocating a run from an old merging run, and then releasing the old merging run.

dcoutts reviewed (approved):
Looking good!

mheinzel commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -327,10 +353,170 @@ remainingMergeDebtPendingMerge (PendingMerge _ prs mts) = do
+    if c0 <= 0
+      then return 0
+      else isStructurallyEmpty mt0 >>= \case
+        True  -> return 0
```
> Is that right? Don't we return the leftover credits rather than 0?

No, it isn't. Also just found this when removing the "expect failure" from the class-based property test.

> Do we need this special case at all? Won't it work out properly with the general case?

No, because we then have an empty list of runs that we need to turn into something. Normally, we feed these runs into `MergingRun.new`, but that errors on an empty list of inputs. So if we don't handle it here, we need to add a special case further down to fabricate an empty run out of thin air.

However, I found that how we handle it here is not right. A structurally empty tree has debt 1, so we need to reduce/remove it somehow when supplying credits (or add another special case in `remainingUnionDebt`).

IMO, when supplying credits to a structurally empty union level, we should immediately remove that union level. But that's a bit awkward here: What should this function do? Signal in the result that the caller should remove the union level? Ideally we should not have created an empty (but existing) union level in the first place.

For almost every union operation I work on I find that I need to add a special case for structurally empty trees. They weaken the invariants/guarantees just to have an obviously empty placeholder value, which bites us in other places. What do we really need empty merging trees for?

mheinzel replied:
I tried handling the structurally empty case further down in `startPendingMerge`, but then `splitEqually` also failed with division by zero, so I moved it to `supplyState` now.

I still think it would simplify the code if we got rid of the structurally empty. Restoring a snapshot probably could fail then or would need some tweaking to propagate the `Maybe`, but I think overall the code would get better. Maybe something to try a bit later.

jorisdral reviewed (commented):
LGTM, modulo some small suggestions

The only thing I haven't looked at super carefully is `MergingTree.supplyCredits`, but it should be fine since @dcoutts hopefully did :stuck_out_tongue_winking_eye:

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -147,17 +147,36 @@ new ::
+            (SpentCredits 0)
+            MergeKnownCompleted
+            (CompletedMerge r)
+        1 -> do
+          r <- dupRun (V.head nonEmptyRuns)
+          unsafeNew
+            (MergeDebt 0)
+            (SpentCredits 0)
+            MergeKnownCompleted
+            (CompletedMerge r)
```
Do the run params of an empty run really matter anywhere? If they do, we can create a new empty run, but that means creating files, so there is some overhead.

mheinzel replied:
I undid it for now.

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -740,6 +741,25 @@ atomicSpendCredits (CreditsVar var) spend =
 
+{-# SPECIALISE remainingMergeDebt ::
+     Ref (MergingRun t IO h) -> IO (MergeDebt, NumEntries) #-}
+-- | Calculate an upper bound on the merge credits required to complete the
+-- merge, as well as an upper bound on the size of the resulting run.
```
In other places we definitely don't have a tight bound, we'll generally overestimate. And here, it's pretty much tight, but not quite clear with concurrency. So I don't think we need to change anything.

jorisdral replied:
My understanding is that the bound is tight (barring concurrency) in the sense that if it returns a remaining debt of `x` then I can *definitely* finish the union/merge if I supply `x` credits after. There is always the possibility that union/merge is not finished if I supply `x-1`. 

So I was just musing whether we could make that more precise in the comment. *An* upper bound kind of sounds like it could generally be something way off as well, such as `maxBound`

mheinzel commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -272,3 +295,42 @@ finalise mergeState = releaseMTS =<< readMVar mergeState
+debtOfNestedMerge debts =
+    -- complete all children, then one merge of them all (so debt is their size)
+    -- and 1 to make sure a pending merge never has debt 0
+    (MergeDebt (c + MR.MergeCredits n + 1), NumEntries n)
```
(already outdated)

mheinzel commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -272,3 +295,42 @@ finalise mergeState = releaseMTS =<< readMVar mergeState
+    -- and 1 to make sure a pending merge never has debt 0
+    (MergeDebt (c + MR.MergeCredits n + 1), NumEntries n)
+  where
+    (MergeDebt c, NumEntries n) = foldl' add (MergeDebt 0, NumEntries 0) debts
+
+    add (MergeDebt !d1, NumEntries !n1) (MergeDebt !d2, NumEntries !n2) =
+        (MergeDebt (d1 + d2), NumEntries (n1 + n2))
```
Not sure if makes a difference either, but the list is usually short. We can try that with the other optimisations once we have a benchmark.

mheinzel commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -272,3 +295,42 @@ finalise mergeState = releaseMTS =<< readMVar mergeState
+remainingMergeDebt (DeRef mt) = do
+    readMVar (mergeState mt) >>= \case
+      CompletedTreeMerge r -> return (MergeDebt 0, Run.size r)
+      OngoingTreeMerge mr  -> MR.remainingMergeDebt mr
```
(outdated)

mheinzel commented on src/Database/LSMTree/Internal.hs:
```diff
@@ -1626,14 +1627,19 @@ newtype UnionDebt = UnionDebt Int
+          NoUnion ->
+            pure (UnionDebt 0)
+          Union mt -> do
+            (MergeDebt (MergeCredits c), _) <- MT.remainingMergeDebt mt
```
I have

```haskell
-- | Calculate an upper bound on the merge credits required to complete the
-- merge, i.e. turn the tree into a 'CompletedTreeMerge'. For the recursive
-- calculation, we also return an upper bound on the size of the resulting run.
remainingMergeDebt ::
     (MonadMVar m, PrimMonad m)
  => Ref (MergingTree m h) -> m (MergeDebt, NumEntries)
```
and later `-- complete all children, then one merge of them all (so debt is their size)`

Maybe there's a better place to put it, e.g. defining a datatype for debt+size and explaining it there.

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -760,6 +762,36 @@ remainingMergeDebt (DeRef mr) = do
+#else
+    supply x credits
+#endif
+  where
+    assertM :: HasCallStack => Bool -> m ()
+    assertM p = let _ = callStack in assert p (pure ())
+    -- just uses callStack so the constraint is not redundant in release builds
```
Without it, if any of the assertions fail, the error doesn't show which one it is. Now we get a helpful line number at least.

mheinzel replied:
> We use this same trick in the `Control.Refcount` module

I tried that, but then I got a warning for `assertM` itself. :(
https://github.com/IntersectMBO/lsm-tree/pull/633#discussion_r2003096591

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -760,6 +762,36 @@ remainingMergeDebt (DeRef mr) = do
+  -> (r -> MergeCredits -> m MergeCredits)  -- how to supply
+  -> (r -> MergeCredits -> m MergeCredits)
+supplyChecked _query supply x credits = do
+    assertM $ credits > 0   -- only call them when there are credits to spend
```
That's a little more precise, true.

mheinzel commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -306,8 +326,17 @@ remainingMergeDebt ::
       CompletedTreeMerge r -> return (MergeDebt 0, Run.size r)
-      OngoingTreeMerge mr  -> MR.remainingMergeDebt mr
-      PendingTreeMerge ptm -> remainingMergeDebtPendingMerge ptm
+      OngoingTreeMerge mr  -> addDebtOne <$> MR.remainingMergeDebt mr
+      PendingTreeMerge ptm -> addDebtOne <$> remainingMergeDebtPendingMerge ptm
+  where
+    -- An ongoing merge should never have 0 debt, even if the 'MergingRun' in it
+    -- says it is completed. We still need to update it to 'CompletedTreeMerge'.
+    -- Similarly, a pending merge needs some work to complete it, even if all
+    -- its inputs are empty.
+    --
+    -- Note that we can't use @max 1@, as this would violate the property that
+    -- supplying N credits reduces the remaining debt by at least N.
+    addDebtOne (MergeDebt !debt, !size) = (MergeDebt (debt + 1), size)
```
I think I squashed it into the wrong commit. Fixed!

mheinzel commented on src/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -327,10 +356,188 @@ remainingMergeDebtPendingMerge (PendingMerge _ prs mts) = do
+                -- all work is done, we can't spend any more credits
+                return (CompletedTreeMerge r, leftovers)
+
+          PendingTreeMerge _
+            | isStructurallyEmptyState state -> do
+            -- make a completely fresh empty run. this can only happen at the
+            -- root. the structurally empty tree still has debt 1, so we want to
+            -- merge it into a single run.
+            -- we handle this as a special case here since in several places
+            -- below we require the list of children to be non-empty.
+            runPaths <- mkFreshRunPaths
+            run <-
+              withRollback reg
+                -- TODO: the builder's handles aren't cleaned up if we fail
+                -- before fromBuilder closes them
+                (Run.fromBuilder =<<
+                   RunBuilder.new hfs hbio runParams runPaths (NumEntries 0))
+                releaseRef
+            return (CompletedTreeMerge run, credits)
```
This also goes away if we don't allow trees to be structurally empty, but yes: If we make runs non-empty, we probably want `CompletedTreeMergeEmpty`.

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -147,17 +147,36 @@ new ::
   -> m (Ref (MergingRun t m h))
 new hfs hbio resolve runParams ty runPaths inputRuns =
+    assert (V.length inputRuns > 0) $ do
+    -- there can be empty runs, which we don't want to include in the merge
+    let nonEmptyRuns = V.filter (\r -> Run.size r > NumEntries 0) inputRuns
```
I left a TODO.

mheinzel commented:
I added some new tests and addressed most feedback. Some comments are still open, I can quickly make the corresponding changes based on what you think. Also feel free to merge if you think it's good enough. I can still make small improvements in a new PR.

jorisdral commented on test/Test/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -178,24 +188,67 @@ modelFoldMergingTree = goMergingTree
+      return (conjoin (toList props))
+  where
+    root = Paths.SessionRoot (FS.mkFsPath [])
+    setupPath = FS.mkFsPath ["setup"]  -- separate dir, so it doesn't clash
```
Will it clash if `withMergingTree` and `supplyCredits` use the same `counter`?

jorisdral commented on test/Test/Database/LSMTree/Internal/MergingTree.hs:
```diff
@@ -178,24 +188,67 @@ modelFoldMergingTree = goMergingTree
+            , counterexample "did not reduce debt sufficiently" $
+                debt' <= debt - (c - leftovers)
+            ]
+      return (conjoin (toList props))
```
It might be nice to label test cases with the final debt, so that we can see whether tree merges ever complete
